pkg/clients/ecr/repository_policy.go

```diff
@@ -209,7 +220,19 @@ func IsRepositoryPolicyUpToDate(local, remote *string) bool {
 		return false
 	}
 
-	return cmp.Equal(localUnmarshalled, remoteUnmarshalled)
+	sortSlicesOpt := cmpopts.SortSlices(func(x, y interface{}) bool {
+		if a, ok := x.(string); ok {
+			if b, ok := y.(string); ok {
+				return a < b
+			}
+		}
+		// Note: Unknown types in slices will not cause a panic, but
+		// may not be sorted correctly. Depending on how AWS handles
+		// these, it may cause constant updates - but better this than
+		// panicing.
+		return false
+	})
```

Comment:
```suggestion
	sortSlicesOpt := cmpopts.SortSlices(func(x, y string) bool { return x < y })
```

I think cmpopts checks the signature and types here. https://play.golang.org/p/RlhDKTVo0-T

Reply:
Mentioned in thread in slack but this behaves differently, looks like it aborts sorting the entire slice as soon as it comes across a value that isn't a string.

Don't know what the correct behaviour here is, if we're happy that slices containing non-strings will be left (partially?) unsorted then I'm happy to change this, it just means in cases where unknown types were already in the right position in the slice we might end up returning `false` instead of `true` and causing an update.

Reply:
I'm happy with the code as it is, I assumed it would be string only because the schema is string only. As long as we end up with one implementation of "compare policy" in the code base I think it's OK if it is very "defensively coded". But if we get 10 copies I think it would be best to handle only cases we know can happen :)

Reply:
If we expect non-string types, would it make sense to supply a sort option for each type we expect? Something like the following:
```go
sortSlicesOpts := cmp.Options{
		cmpopts.SortSlices(func(x, y string) bool { return x < y }),
		cmpopts.SortSlices(func(x, y int) bool { return x < y }),
		cmpopts.SortSlices(func(x, y bool) bool { return x != y }),
	}
return cmp.Equal(localUnmarshalled, remoteUnmarshalled, cmpopts.EquateEmpty(), sortSlicesOpts)
```

I don't know how `cmp` behaves with JSON though, so might be totally off base here.

Reply:
@muvaf According to all of the doco I've seen, and the policy grammar that @chlunde found (https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_grammar.html) it seems that _everything_ should be a string, including policy conditions that contain numbers or booleans - so I don't think we necessarily have to sort non-string types like that, as they may well be completely rejected by AWS anyway.

We do however need to _accept_ them without error, as a user can supply whatever JSON document they want.

Rather than adding additional types which I think will probably be rejected by the API anyway, would it be better if we simply provide an `interface{}` sorter that always returns false, alongside the basic string sorter? That way, we should always complete sorting the input _strings_ even if we encounter other types, and let the API deal with the fallout.

```go
sortSlicesOpts := cmp.Options{
	cmpopts.SortSlices(func(x, y string) bool { return x < y }),
	cmpopts.SortSlices(func(x, y interface{}) bool { return false }),
}
return cmp.Equal(localUnmarshalled, remoteUnmarshalled, cmpopts.EquateEmpty(), sortSlicesOpts)
```

Reply:
Sounds good to me. Let me know when it's ready to merge.

Reply:
I've done a load of testing on this and the above code sample doesn't work - and neither does @chlunde's original suggestion. I suspect this is because the values as seen by `cmp.Equal` are never directly strings, but strings inside interfaces, because they're generically unmarshalled - so we can't detect the variable type as part of the options sort.

The existing code in this PR _does_ do what we want but is a little more verbose because of this limitation.

Reply:
@benagricola sorry for the extra work, I think you're right. `cmpopts` doesn't check the types of each element when we have `[]interface{}`: https://play.golang.org/p/XdzfTqK_idZ - so most likely we have a `[]interface{}` and not `[]string`